server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustLanguagePlugin.kt

```diff
@@ -0,0 +1,331 @@
+                    RustTarget(
+                        resolvePackage(genericData).targetName,
+                        rustData.crateRoot,
+                        genericData.baseDirectory.toString(),
+                        rustData.kind,
+                        rustData.edition,
+                        false,                                  // TODO: check it somehow. I even know where to look for it :/  http://bazelbuild.github.io/rules_rust/rust_doc.html
+                        rustData.crateFeatures
+                    )
+                }
+
+                val procMacroArtifacts = rustTargetsWithData
+                    .flatMap { (_, rustData) -> rustData.procMacroArtifacts }
+                    .map { bazelPathsResolver.pathToDirectoryUri(it) }
+                    .map { RustProcMacroArtifact(it.path, "") }
+
+                val procMacroArtifact = procMacroArtifacts.firstOrNull() // TODO: oh yeah, we need toolchain here...
+
+                val targets = rustTargetsWithData.map(mapTarget)
+                val allTargets = allRustTargetsWithData.map(mapTarget)
+                val allFeatures = rustTargetsWithData.flatMap { (_, rustData) ->
+                    rustData.crateFeatures
+                }
+                RustPackage(
+                    rustPackage,
+                    version,
+                    origin,
+                    edition,
+                    source,
+                    targets,
+                    allTargets,
+                    allFeatures.map { RustFeature(it, listOf()) },
+                    allFeatures,
+                    null,
+                    listOf(
+                        RustEnvData("CARGO_MANIFEST_DIF", "$pkgBaseDir${rustPackage.drop(1)}"),
+                        RustEnvData("CARGO", "cargo"),
+                        RustEnvData("CARGO_PKG_VERSION", version),
+                        RustEnvData("CARGO_PKG_VERSION_MAJOR", major),
+                        RustEnvData("CARGO_PKG_VERSION_MINOR", minor),
+                        RustEnvData("CARGO_PKG_VERSION_PATCH", patch),
+                        RustEnvData("CARGO_PKG_VERSION_PRE", ""),
+                        RustEnvData("CARGO_PKG_AUTHORS", ""),
+                        RustEnvData("CARGO_PKG_NAME", rustPackage),
+                        RustEnvData("CARGO_PKG_DESCRIPTION", ""),
+                        RustEnvData("CARGO_PKG_REPOSITORY", ""),
+                        RustEnvData("CARGO_PKG_LICENSE", ""),
+                        RustEnvData("CARGO_PKG_LICENSE_FILE", ""),
+                        RustEnvData("CARGO_CRATE_NAME", rustPackage),
+                    ),
+                    null,
+                    procMacroArtifact
+                )
+            }
+
+        return packages
+    }
+
+    // We need to resolve all dependencies are provide a list of new bazel targets (deps) to be transformed into packages.
+    private fun rustDependencies(
```

Comment:
imo it'd be nice to move it to a separate class and write some tests

Reply:
@Perlik476 this looks like a job for you